src/Analyzer.Core.JsonRuleTests/TestRunner.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.Templates.Analyzer.Types;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Templates.Analyzer.Core.JsonRuleTests
+{
+    [TestClass]
+    public class TestRunner
+    {
+        Dictionary<string, IEnumerable<IEvaluation>> templateEvaluations = new();
+
+        /// <summary>
+        /// Runs a test defined in a test configuration file in the Tests directory.
+        /// </summary>
+        /// <param name="ruleExpectations">The test configuration to run.</param>
+        [DataTestMethod]
+        [DynamicData(nameof(GetTests), DynamicDataSourceType.Method, DynamicDataDisplayName = nameof(GetTestDisplayName))]
+        public void TestRule(TestConfiguration ruleExpectations)
+        {
+            // Get template to analyze
+            var testTemplatePath = Path.Combine("TestTemplates", $"{ruleExpectations.Template}.badtemplate");
+            var testTemplate = File.ReadAllText(testTemplatePath);
+
+            // If not already analyzed, analyze it and store evaluations
+            if (!templateEvaluations.TryGetValue(testTemplatePath, out var results))
+            {
+                var templateAnalyzer = new TemplateAnalyzer(testTemplate);
+                results = templateAnalyzer.EvaluateRulesAgainstTemplate();
+                templateEvaluations[testTemplatePath] = results;
+            }
+
+            // Find any instances of the rule being tested
+            var thisRuleEvaluations = results.ToList().Where(e => e.RuleName.Equals(ruleExpectations.TestName, StringComparison.OrdinalIgnoreCase)).ToList();
+
+            if (thisRuleEvaluations.Count == 0)
+            {
+                // Rule should be expected to pass if no results are returned
+                Assert.IsTrue(ruleExpectations.ExpectPass);
+            }
+
+            foreach (var evaluation in thisRuleEvaluations)
+            {
+                Assert.AreEqual(ruleExpectations.ExpectPass, evaluation.Passed);
+
+                if (!evaluation.Passed)
+                {
+                    // Get all lines reported as failed
+                    var failingLines = GetAllResults(evaluation)
+                        .Where(r => !r.Passed)
+                        .Select(r => r.LineNumber).ToHashSet();
+
+                    // Verify all expected lines are reported
+                    var expectedLines = ruleExpectations.ReportedLines.ToHashSet();
+                    Assert.IsTrue(failingLines.SetEquals(expectedLines),
+                        "Expected failing lines do not match actual failed lines.  " +
```

Comment:
I did not know you can add a message to Assert.IsTrue. Cool! This is useful for all the tests we do DeepEquals on JSON objects.

Reply:
Pretty much all the functions in the `Assert` class have an optional message argument :)